# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: 

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [5]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [6]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [7]:
inventories.head()

id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1

In [8]:
inventories[inventories['set_num'] == '663-1']

id  version set_num
10264  15865        1   663-1

In [9]:
inventory_sets.head()

inventory_id  set_num  quantity
0            35  75911-1         1
1            35  75912-1         1
2            39  75048-1         1
3            39  75053-1         1
4            50   4515-1         1

In [10]:
parts.head()

part_num                                        name  part_cat_id  \
0   003381                 Sticker Sheet for Set 663-1           58   
1   003383         Sticker Sheet for Sets 618-1, 628-2           58   
2   003402  Sticker Sheet for Sets 310-3, 311-1, 312-3           58   
3   003429                Sticker Sheet for Set 1550-1           58   
4   003432  Sticker Sheet for Sets 357-1, 355-1, 940-1           58   

  part_material  
0       Plastic  
1       Plastic  
2       Plastic  
3       Plastic  
4       Plastic

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [11]:
def data_count(df):
    return len(df.index)

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [12]:
Q1 = data_count(sets)
Q1

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [13]:
sets

set_num                             name  year  theme_id  num_parts
0            001-1                            Gears  1965         1         43
1           0011-2                Town Mini-Figures  1979        67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199          0
3           0012-1               Space Mini-Figures  1979       143         12
4           0013-1               Space Mini-Figures  1979       143         12
...            ...                              ...   ...       ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019       158         60
18572      XWING-2                X-Wing Trench Run  2019       158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158        413
18574   YTERRIER-1                Yorkshire Terrier  2018       598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501          0

[18576 rows x 5 columns]

In [14]:
Q2 = round(sets['num_parts'].mean())
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [15]:
Q3 = round(sets['num_parts'].median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [16]:
Q4 = round(sets['num_parts'].max())
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [17]:
Q5 = sets[sets['num_parts']  == 11695]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [18]:
def select_set_row(set_num):
    return sets[sets['set_num']  == set_num]
   

In [19]:
#check
select_set_row('71040-1')

set_num           name  year  theme_id  num_parts
12144  71040-1  Disney Castle  2016       608       4081

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [20]:
def select_set_numparts(num):
    return sets[sets['num_parts']  == num].sort_values(by=['year', 'set_num'])
    

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [21]:
select_set_numparts(0)

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [22]:
Q8A = sets.year.min()
Q8A

1949

In [23]:
Q8B = sets.year.max()
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.
- Use the above function to return a DataFrame for rows that match the minimum year (calculated as `Q8A` above) and call this `Q9A`.  
- Then use the above function to return a DataFrame for rows that match the maximum year (calculated as `Q8B` above) and call this `Q9B`.

In [24]:
def select_set_year(num):
    return sets[sets['year']  == num].sort_values(by='set_num')

In [25]:
Q9A = select_set_year(1949)
Q9A

set_num                               name  year  theme_id  num_parts
11057  700.1-1         Extra-Large Gift Set (ABB)  1949       365        142
11070  700.2-1               Large Gift Set (ABB)  1949       365        178
11075  700.3-1              Medium Gift Set (ABB)  1949       365        142
11080  700.A-1              Small Brick Set (ABB)  1949       371         24
11081  700.B-1  Small Doors and Windows Set (ABB)  1949       371         12

In [26]:
Q9B = select_set_year(2022)
Q9B

set_num                                               name  \
336            10297-1                                     Boutique Hotel   
600            10781-1                          Spider-Man’s Techno Trike   
601            10782-1                      Hulk vs. Rhino Truck Showdown   
602            10783-1                        Spider-Man at Doc Ock’s Lab   
603            10784-1                     Spider-Man Webquarters Hangout   
...                ...                                                ...   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   
17207  9780794449247-1  DC Comics Super Heroes: Gotham City’s New Defe...   
17726    DOTS-SERIES-6                    DOTS Collectible Tiles Series 6   

       year  theme_id  num_parts  
336    2022       155       3068  
600    2022       706         59  
601    2022       706        110  
602    2022       706        131  
603    2022       706        155  
...     ...       ...        ...  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  
17207  2022       497          7  
17726  2022       688         16  

[183 rows x 5 columns]

**Q10:** What is the median year represented in the `sets` data?  Save this as `Q10` (as an integer).

In [27]:
Q10 = round(sets['year'].median())
Q10

2010

Notice that the first set was back in 1949 but over half of the sets were after the median year. This could be an indication that the number of sets has increased greatly over the past number of years.

**Q11:** What is the average number of sets listed per year?  Save this as `Q11` (as an integer).

In [28]:
Q11 = round(sets.groupby('year')['set_num'].count().mean())
Q11

258

**Q12:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [29]:
#look at df for refresh
sets

set_num                             name  year  theme_id  num_parts
0            001-1                            Gears  1965         1         43
1           0011-2                Town Mini-Figures  1979        67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199          0
3           0012-1               Space Mini-Figures  1979       143         12
4           0013-1               Space Mini-Figures  1979       143         12
...            ...                              ...   ...       ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019       158         60
18572      XWING-2                X-Wing Trench Run  2019       158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158        413
18574   YTERRIER-1                Yorkshire Terrier  2018       598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501          0

[18576 rows x 5 columns]

In [30]:
themes

id                        name  parent_id
0      1                     Technic        NaN
1      3                 Competition        1.0
2      4              Expert Builder        1.0
3     16                  RoboRiders        1.0
4     17              Speed Slammers        1.0
..   ...                         ...        ...
439  715                      Marvel      535.0
440  716                     Modulex        NaN
441  717                 Speed Racer        NaN
442  718       Series 22 Minifigures      535.0
443  719  BrickLink Designer Program        NaN

[444 rows x 3 columns]

In [31]:
df_joined = sets.merge(themes, left_on='theme_id', right_on='id', how='left', suffixes=('_sets','_themes'))
#df_joined = df_joined.drop_duplicates(subset=['theme_id'])
df_joined

set_num                        name_sets  year  theme_id  \
0            001-1                            Gears  1965         1   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18571      XWING-1              Mini X-Wing Fighter  2019       158   
18572      XWING-2                X-Wing Trench Run  2019       158   
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18574   YTERRIER-1                Yorkshire Terrier  2018       598   
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501   

       num_parts   id   name_themes  parent_id  
0             43    1       Technic        NaN  
1             12   67  Classic Town       50.0  
2              0  199  Lion Knights      186.0  
3             12  143  Supplemental      126.0  
4             12  143  Supplemental      126.0  
...          ...  ...           ...        ...  
18571         60  158     Star Wars        NaN  
18572         52  158     Star Wars        NaN  
18573        413  158     Star Wars        NaN  
18574          0  598   Promotional        NaN  
18575          0  501          Gear        NaN  

[18576 rows x 8 columns]

In [32]:
def theme_by_year(num):
    df = df_joined[df_joined['year'] == num]
    return df[['id', 'name_themes']].sort_values(by='id').drop_duplicates(subset='id').reset_index(drop=True)

In [33]:
theme_by_year(1960)

id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic

In [34]:
theme_by_year(1978)

id               name_themes
0     4            Expert Builder
1    67              Classic Town
2   189            Classic Castle
3   227                 Christmas
4   235                      4.5V
5   363                      Boat
6   364  Building Set with People
7   404                Hobby Sets
8   405                 Homemaker
9   412                   Airport
10  453                   Technic
11  469                 Basic Set
12  473              Supplemental
13  501                      Gear
14  505                 Basic Set
15  632                      Town

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [35]:
theme_by_year(1991)

id             name_themes
0     1                 Technic
1    19            Supplemental
2    21  Universal Building Set
3    67            Classic Town
4   129            Blacktron II
5   136                  M:Tron
6   143            Supplemental
7   148               Pirates I
8   151       Imperial Soldiers
9   190               Crusaders
10  227               Christmas
11  236                      9V
12  278              Model Team
13  363                    Boat
14  443           Service Packs
15  453                 Technic
16  456                   Train
17  467                   Basic
18  469               Basic Set
19  471                 Ferries
20  473            Supplemental
21  498                 Technic
22  504                   Duplo
23  505               Basic Set
24  516       Duplo and Explore
25  528            Supplemental
26  533                    Town
27  534  Universal Building Set
28  559             Value Packs
29  632                    Town
30  634                  Trains
31  652               Playhouse

**Q13:** What theme has the most number of sets associated with it?  Save the theme name as `Q13` (as a string).

In [36]:
df_joined.groupby('name_themes')[['set_num']].count().sort_values(by='set_num', ascending=False).head(1)

set_num
name_themes         
Gear            1904

In [37]:
Q13 = 'Gear'
Q13

'Gear'

**Q14:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

## need to come back!!!!!

In [38]:
df_joined2 = sets.merge(themes, left_on='theme_id', right_on='id', suffixes=('_sets','_themes'))

In [39]:
def theme_by_name(name):
    df = df_joined2[df_joined2['name_themes'] == name]
    return df[['set_num', 'name_sets', 'year', 'num_parts', 'theme_id', 'name_themes']].sort_values(by='set_num').reset_index(drop=True)

**Code Check:** Call the `theme_by_name()` function using the theme name from `Q13`.  You should return 1,904 sets that match that theme name.

In [40]:
theme_by_name('Supplemental')

set_num                           name_sets  year  num_parts  theme_id  \
0    0012-1                  Space Mini-Figures  1979         12       143   
1    0013-1                  Space Mini-Figures  1979         12       143   
2    0014-1                  Space Mini-Figures  1979         12       143   
3    0015-1                  Space Mini-Figures  1979         18       143   
4     045-2      Assorted Clear Bricks & Plates  1969         37       371   
..      ...                                 ...   ...        ...       ...   
435  9919-1        X-Large Storage Bin - 3 Pack  1997          6       528   
436  9920-1          X-Large Yellow Storage Bin  1990          2       528   
437  9921-1                   Large Storage Bin  1995          2       528   
438   994-1       LEGO fences with garden gates  1969         19       371   
439   995-1  Lighting Bricks with Color Filters  1969         23       371   

      name_themes  
0    Supplemental  
1    Supplemental  
2    Supplemental  
3    Supplemental  
4    Supplemental  
..            ...  
435  Supplemental  
436  Supplemental  
437  Supplemental  
438  Supplemental  
439  Supplemental  

[440 rows x 6 columns]

**Q15:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [41]:
df_joined2

set_num                           name_sets  year  theme_id  \
0            001-1                               Gears  1965         1   
1            002-1      4.5V Samsonite Gears Motor Set  1965         1   
2           1030-1      TECHNIC I: Simple Machines Set  1985         1   
3           1038-1                  ERBIE the Robo-Car  1985         1   
4           1039-1                Manual Control Set 1  1986         1   
...            ...                                 ...   ...       ...   
18571  M20-2566-10  Modulex Box - 10x Window 1 x 6 x 6  1963       716   
18572  M20-2575-10  Modulex Box - 10x Window 1 x 7 x 5  1963       716   
18573  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963       716   
18574  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963       716   
18575  WEETABIX1-1                     Weetabix Castle  1970       414   

       num_parts   id name_themes  parent_id  
0             43    1     Technic        NaN  
1              3    1     Technic        NaN  
2            211    1     Technic        NaN  
3            120    1     Technic        NaN  
4             39    1     Technic        NaN  
...          ...  ...         ...        ...  
18571         10  716     Modulex        NaN  
18572         10  716     Modulex        NaN  
18573         10  716     Modulex        NaN  
18574         10  716     Modulex        NaN  
18575        471  414      Castle      411.0  

[18576 rows x 8 columns]

In [42]:
def theme_by_setnum(set_num):
    df = df_joined2[df_joined2['set_num'] == set_num]
    return df['name_themes'].values[0]
    

In [43]:
theme_by_setnum('1039-1')

'Technic'

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [44]:
theme_by_setnum('11002-1')

'Classic'

**Q16:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You **do not** need to save this as `Q16`.  Instead you will use this information for later questions.

In [45]:
df_joined2[df_joined2['name_sets'] == 'Millennium Falcon'].sort_values(by='num_parts', ascending=False).head(1)

set_num          name_sets  year  theme_id  num_parts   id  \
3199  75192-1  Millennium Falcon  2017       171       7541  171   

                    name_themes  parent_id  
3199  Ultimate Collector Series      158.0

**Q17:** 
- How many minifigs are contained in the set from `Q16`?  Save this as `Q17` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [46]:
sets

set_num                             name  year  theme_id  num_parts
0            001-1                            Gears  1965         1         43
1           0011-2                Town Mini-Figures  1979        67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199          0
3           0012-1               Space Mini-Figures  1979       143         12
4           0013-1               Space Mini-Figures  1979       143         12
...            ...                              ...   ...       ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019       158         60
18572      XWING-2                X-Wing Trench Run  2019       158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158        413
18574   YTERRIER-1                Yorkshire Terrier  2018       598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501          0

[18576 rows x 5 columns]

In [47]:
inventories

id  version        set_num
0           1        1         7922-1
1           3        1         3931-1
2           4        1         6942-1
3          15        1         5158-1
4          16        1          903-1
...       ...      ...            ...
31582  122328        2          426-1
31583  122329        1  210-2-DUMMY-1
31584  122344        2         7036-1
31585  122450        1        11961-1
31586  122472        3        10262-1

[31587 rows x 3 columns]

In [48]:
inventory_minifigs

inventory_id     fig_num  quantity
0                 3  fig-001549         1
1                 4  fig-000764         1
2                19  fig-000555         1
3                25  fig-000574         1
4                26  fig-000842         1
...             ...         ...       ...
18139        122344  fig-009900         1
18140        122344  fig-009908         2
18141        122344  fig-009954         1
18142        122344  fig-009955         1
18143        122344  fig-009956         1

[18144 rows x 3 columns]

In [49]:
df_joined3 = sets.merge(inventories, left_on='set_num', right_on='set_num')
df_joined4 = df_joined3.merge(inventory_minifigs, left_on='id', right_on='inventory_id')
df_joined4

set_num                             name  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts     id  version  inventory_id     fig_num  quantity  
0             12   5087        1          5087  fig-000225         1  
1             12   5087        1          5087  fig-000834         1  
2             12   5087        1          5087  fig-004401         1  
3             12   1414        1          1414  fig-000020         2  
4             12   4609        1          4609  fig-001127         2  
...          ...    ...      ...           ...         ...       ...  
18139        147  13813        1         13813  fig-001361         1  
18140          6  34919        1         34919  fig-010738         1  
18141        413  36159        1         36159  fig-004483         1  
18142        413  36159        1         36159  fig-004484         1  
18143        413  36159        1         36159  fig-010456         1  

[18144 rows x 10 columns]

In [50]:
df_falc = df_joined4[df_joined4['set_num'] == '75192-1']
df_falc

set_num               name  year  theme_id  num_parts     id  version  \
12745  75192-1  Millennium Falcon  2017       171       7541  19670        1   
12746  75192-1  Millennium Falcon  2017       171       7541  19670        1   
12747  75192-1  Millennium Falcon  2017       171       7541  19670        1   
12748  75192-1  Millennium Falcon  2017       171       7541  19670        1   
12749  75192-1  Millennium Falcon  2017       171       7541  19670        1   
12750  75192-1  Millennium Falcon  2017       171       7541  19670        1   
12751  75192-1  Millennium Falcon  2017       171       7541  19670        1   
12752  75192-1  Millennium Falcon  2017       171       7541  19670        1   

       inventory_id     fig_num  quantity  
12745         19670  fig-001714         1  
12746         19670  fig-001810         1  
12747         19670  fig-002057         1  
12748         19670  fig-002514         1  
12749         19670  fig-002544         1  
12750         19670  fig-002549         1  
12751         19670  fig-002550         1  
12752         19670  fig-002551         1

In [51]:
Q17 = df_falc['quantity'].sum()
Q17

8

**Q18:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [52]:
df_joined4

set_num                             name  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts     id  version  inventory_id     fig_num  quantity  
0             12   5087        1          5087  fig-000225         1  
1             12   5087        1          5087  fig-000834         1  
2             12   5087        1          5087  fig-004401         1  
3             12   1414        1          1414  fig-000020         2  
4             12   4609        1          4609  fig-001127         2  
...          ...    ...      ...           ...         ...       ...  
18139        147  13813        1         13813  fig-001361         1  
18140          6  34919        1         34919  fig-010738         1  
18141        413  36159        1         36159  fig-004483         1  
18142        413  36159        1         36159  fig-004484         1  
18143        413  36159        1         36159  fig-010456         1  

[18144 rows x 10 columns]

In [53]:
#df_joined4.rename({'name': 'fig_name'}, axis=1, inplace=True)

In [54]:
inventory_minifigs

inventory_id     fig_num  quantity
0                 3  fig-001549         1
1                 4  fig-000764         1
2                19  fig-000555         1
3                25  fig-000574         1
4                26  fig-000842         1
...             ...         ...       ...
18139        122344  fig-009900         1
18140        122344  fig-009908         2
18141        122344  fig-009954         1
18142        122344  fig-009955         1
18143        122344  fig-009956         1

[18144 rows x 3 columns]

In [55]:
inventories

id  version        set_num
0           1        1         7922-1
1           3        1         3931-1
2           4        1         6942-1
3          15        1         5158-1
4          16        1          903-1
...       ...      ...            ...
31582  122328        2          426-1
31583  122329        1  210-2-DUMMY-1
31584  122344        2         7036-1
31585  122450        1        11961-1
31586  122472        3        10262-1

[31587 rows x 3 columns]

In [56]:
df_joined5 = df_joined4.merge(inventories, left_on='id', right_on='id', suffixes=('_df','_fig'))
df_joined5

set_num_df                             name  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts     id  version_df  inventory_id     fig_num  quantity  \
0             12   5087           1          5087  fig-000225         1   
1             12   5087           1          5087  fig-000834         1   
2             12   5087           1          5087  fig-004401         1   
3             12   1414           1          1414  fig-000020         2   
4             12   4609           1          4609  fig-001127         2   
...          ...    ...         ...           ...         ...       ...   
18139        147  13813           1         13813  fig-001361         1   
18140          6  34919           1         34919  fig-010738         1   
18141        413  36159           1         36159  fig-004483         1   
18142        413  36159           1         36159  fig-004484         1   
18143        413  36159           1         36159  fig-010456         1   

       version_fig  set_num_fig  
0                1       0011-2  
1                1       0011-2  
2                1       0011-2  
3                1       0012-1  
4                1       0013-1  
...            ...          ...  
18139            1  WEETABIX4-1  
18140            1    WILLIAM-1  
18141            1  YODACHRON-1  
18142            1  YODACHRON-1  
18143            1  YODACHRON-1  

[18144 rows x 12 columns]

In [57]:
df_joined5.rename({'set_num_df': 'set_num'}, axis=1, inplace=True)

In [58]:
df_joined5

set_num                             name  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts     id  version_df  inventory_id     fig_num  quantity  \
0             12   5087           1          5087  fig-000225         1   
1             12   5087           1          5087  fig-000834         1   
2             12   5087           1          5087  fig-004401         1   
3             12   1414           1          1414  fig-000020         2   
4             12   4609           1          4609  fig-001127         2   
...          ...    ...         ...           ...         ...       ...   
18139        147  13813           1         13813  fig-001361         1   
18140          6  34919           1         34919  fig-010738         1   
18141        413  36159           1         36159  fig-004483         1   
18142        413  36159           1         36159  fig-004484         1   
18143        413  36159           1         36159  fig-010456         1   

       version_fig  set_num_fig  
0                1       0011-2  
1                1       0011-2  
2                1       0011-2  
3                1       0012-1  
4                1       0013-1  
...            ...          ...  
18139            1  WEETABIX4-1  
18140            1    WILLIAM-1  
18141            1  YODACHRON-1  
18142            1  YODACHRON-1  
18143            1  YODACHRON-1  

[18144 rows x 12 columns]

In [59]:
minifigs

fig_num                                   name  num_parts
0      fig-000001                     Toy Store Employee          4
1      fig-000002                           Customer Kid          4
2      fig-000003                  Assassin Droid, White          8
3      fig-000004                           Basic Figure          4
4      fig-000005        Captain America with Short Legs          3
...           ...                                    ...        ...
11822  fig-012221                                Bob-omb         11
11823  fig-012222                   Boomer (Banzai) Bill         10
11824  fig-012223                            Bullet Bill         10
11825  fig-012224                           Buzzy Beetle          7
11826  fig-012225  Unikitty - Cheerykitty (Cheery Kitty)         14

[11827 rows x 3 columns]

In [60]:
df_joined6 = df_joined5.merge(minifigs, left_on='fig_num', right_on='fig_num', how='left')
df_joined6

set_num                           name_x  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts_x     id  version_df  inventory_id     fig_num  quantity  \
0               12   5087           1          5087  fig-000225         1   
1               12   5087           1          5087  fig-000834         1   
2               12   5087           1          5087  fig-004401         1   
3               12   1414           1          1414  fig-000020         2   
4               12   4609           1          4609  fig-001127         2   
...            ...    ...         ...           ...         ...       ...   
18139          147  13813           1         13813  fig-001361         1   
18140            6  34919           1         34919  fig-010738         1   
18141          413  36159           1         36159  fig-004483         1   
18142          413  36159           1         36159  fig-004484         1   
18143          413  36159           1         36159  fig-010456         1   

       version_fig  set_num_fig  \
0                1       0011-2   
1                1       0011-2   
2                1       0011-2   
3                1       0012-1   
4                1       0013-1   
...            ...          ...   
18139            1  WEETABIX4-1   
18140            1    WILLIAM-1   
18141            1  YODACHRON-1   
18142            1  YODACHRON-1   
18143            1  YODACHRON-1   

                                                  name_y  num_parts_y  
0      Policeman, Black Suit with Pocket and Badge, W...            4  
1                Man, Red Torso, Blue Legs, Red Hard Hat            4  
2                           Woman - Blue Torso, Red Legs            4  
3      Classic Spaceman, Red with Airtanks (3842a Hel...            5  
4      Classic Spaceman, White with Airtanks (3842a H...            5  
...                                                  ...          ...  
18139         LEGOLAND White Torso, Black Legs, Red Hair            4  
18140                                          Will.i.am            4  
18141  Yoda, Sand Green Head with Gray Hair, Dark Ora...            3  
18142                                    Padawan, Female            4  
18143                                     Holocron Droid           15  

[18144 rows x 14 columns]

In [61]:
df_joined6 = df_joined6.rename({'num_parts_y': 'fig_num_parts'}, axis=1)
df_joined6

set_num                           name_x  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts_x     id  version_df  inventory_id     fig_num  quantity  \
0               12   5087           1          5087  fig-000225         1   
1               12   5087           1          5087  fig-000834         1   
2               12   5087           1          5087  fig-004401         1   
3               12   1414           1          1414  fig-000020         2   
4               12   4609           1          4609  fig-001127         2   
...            ...    ...         ...           ...         ...       ...   
18139          147  13813           1         13813  fig-001361         1   
18140            6  34919           1         34919  fig-010738         1   
18141          413  36159           1         36159  fig-004483         1   
18142          413  36159           1         36159  fig-004484         1   
18143          413  36159           1         36159  fig-010456         1   

       version_fig  set_num_fig  \
0                1       0011-2   
1                1       0011-2   
2                1       0011-2   
3                1       0012-1   
4                1       0013-1   
...            ...          ...   
18139            1  WEETABIX4-1   
18140            1    WILLIAM-1   
18141            1  YODACHRON-1   
18142            1  YODACHRON-1   
18143            1  YODACHRON-1   

                                                  name_y  fig_num_parts  
0      Policeman, Black Suit with Pocket and Badge, W...              4  
1                Man, Red Torso, Blue Legs, Red Hard Hat              4  
2                           Woman - Blue Torso, Red Legs              4  
3      Classic Spaceman, Red with Airtanks (3842a Hel...              5  
4      Classic Spaceman, White with Airtanks (3842a H...              5  
...                                                  ...            ...  
18139         LEGOLAND White Torso, Black Legs, Red Hair              4  
18140                                          Will.i.am              4  
18141  Yoda, Sand Green Head with Gray Hair, Dark Ora...              3  
18142                                    Padawan, Female              4  
18143                                     Holocron Droid             15  

[18144 rows x 14 columns]

In [62]:
df_joined6 = df_joined6.rename({'name_y': 'fig_name'}, axis=1)
df_joined6

set_num                           name_x  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts_x     id  version_df  inventory_id     fig_num  quantity  \
0               12   5087           1          5087  fig-000225         1   
1               12   5087           1          5087  fig-000834         1   
2               12   5087           1          5087  fig-004401         1   
3               12   1414           1          1414  fig-000020         2   
4               12   4609           1          4609  fig-001127         2   
...            ...    ...         ...           ...         ...       ...   
18139          147  13813           1         13813  fig-001361         1   
18140            6  34919           1         34919  fig-010738         1   
18141          413  36159           1         36159  fig-004483         1   
18142          413  36159           1         36159  fig-004484         1   
18143          413  36159           1         36159  fig-010456         1   

       version_fig  set_num_fig  \
0                1       0011-2   
1                1       0011-2   
2                1       0011-2   
3                1       0012-1   
4                1       0013-1   
...            ...          ...   
18139            1  WEETABIX4-1   
18140            1    WILLIAM-1   
18141            1  YODACHRON-1   
18142            1  YODACHRON-1   
18143            1  YODACHRON-1   

                                                fig_name  fig_num_parts  
0      Policeman, Black Suit with Pocket and Badge, W...              4  
1                Man, Red Torso, Blue Legs, Red Hard Hat              4  
2                           Woman - Blue Torso, Red Legs              4  
3      Classic Spaceman, Red with Airtanks (3842a Hel...              5  
4      Classic Spaceman, White with Airtanks (3842a H...              5  
...                                                  ...            ...  
18139         LEGOLAND White Torso, Black Legs, Red Hair              4  
18140                                          Will.i.am              4  
18141  Yoda, Sand Green Head with Gray Hair, Dark Ora...              3  
18142                                    Padawan, Female              4  
18143                                     Holocron Droid             15  

[18144 rows x 14 columns]

In [63]:
def get_minifigs(set_num, version_num):
    df = df_joined6[(df_joined6['set_num'] == set_num)   &  (df_joined6['version_fig'] == version_num)]
    return df[['set_num', 'fig_num', 'fig_name', 'fig_num_parts']].sort_values(by='fig_num').reset_index(drop=True).drop_duplicates()

**Q19:** Using the above function, output information about the minifigs from the set discussed in `Q16`.  Save this output as `Q19`.

In [64]:
#check first before answering this question
df_joined6[df_joined6['set_num'] == '10217-1']

set_num        name_x  year  theme_id  num_parts_x     id  version_df  \
338  10217-1  Diagon Alley  2011       246         2032  14029           1   
339  10217-1  Diagon Alley  2011       246         2032  14029           1   
340  10217-1  Diagon Alley  2011       246         2032  14029           1   
341  10217-1  Diagon Alley  2011       246         2032  14029           1   
342  10217-1  Diagon Alley  2011       246         2032  14029           1   
343  10217-1  Diagon Alley  2011       246         2032  14029           1   
344  10217-1  Diagon Alley  2011       246         2032  14029           1   
345  10217-1  Diagon Alley  2011       246         2032  14029           1   
346  10217-1  Diagon Alley  2011       246         2032  14029           1   
347  10217-1  Diagon Alley  2011       246         2032  14029           1   
348  10217-1  Diagon Alley  2011       246         2032  14029           1   
349  10217-1  Diagon Alley  2011       246         2032  43287           2   
350  10217-1  Diagon Alley  2011       246         2032  43287           2   
351  10217-1  Diagon Alley  2011       246         2032  43287           2   
352  10217-1  Diagon Alley  2011       246         2032  43287           2   
353  10217-1  Diagon Alley  2011       246         2032  43287           2   
354  10217-1  Diagon Alley  2011       246         2032  43287           2   
355  10217-1  Diagon Alley  2011       246         2032  43287           2   
356  10217-1  Diagon Alley  2011       246         2032  43287           2   
357  10217-1  Diagon Alley  2011       246         2032  43287           2   
358  10217-1  Diagon Alley  2011       246         2032  43287           2   
359  10217-1  Diagon Alley  2011       246         2032  43287           2   

     inventory_id     fig_num  quantity  version_fig set_num_fig  \
338         14029  fig-002229         1            1     10217-1   
339         14029  fig-006075         1            1     10217-1   
340         14029  fig-006084         1            1     10217-1   
341         14029  fig-006130         1            1     10217-1   
342         14029  fig-006138         1            1     10217-1   
343         14029  fig-006162         1            1     10217-1   
344         14029  fig-006163         2            1     10217-1   
345         14029  fig-006164         1            1     10217-1   
346         14029  fig-006165         1            1     10217-1   
347         14029  fig-006166         1            1     10217-1   
348         14029  fig-006167         1            1     10217-1   
349         43287  fig-002229         1            2     10217-1   
350         43287  fig-006075         1            2     10217-1   
351         43287  fig-006084         1            2     10217-1   
352         43287  fig-006130         1            2     10217-1   
353         43287  fig-006138         1            2     10217-1   
354         43287  fig-006162         1            2     10217-1   
355         43287  fig-006163         2            2     10217-1   
356         43287  fig-006164         1            2     10217-1   
357         43287  fig-006165         1            2     10217-1   
358         43287  fig-006166         1            2     10217-1   
359         43287  fig-006167         1            2     10217-1   

                                              fig_name  fig_num_parts  
338                   Skeleton (Standard) (3626b Head)              6  
339                 Rubeus Hagrid, Dark Brown Overcoat              3  
340  Lucius Malfoy, Black Robes and Cape, Death Eat...              5  
341  Harry Potter, Open Dark Blue Jacket over White...              4  
342                                    Fenrir Greyback              4  
343  Ron Weasley, Red Tartan Sweater Vest, Reddish ...              4  
344          Fred / George Weasley in Dark Orange Suit              4  
345                                 Garrick Ollivander              4

In [65]:
get_minifigs('10217-1', 1)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

In [66]:
Q19 = get_minifigs('75192-1', 1)
Q19

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [67]:
get_minifigs('10217-1', 1)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

In [68]:
get_minifigs('10217-1', 2)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q20:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [69]:
themes

id                        name  parent_id
0      1                     Technic        NaN
1      3                 Competition        1.0
2      4              Expert Builder        1.0
3     16                  RoboRiders        1.0
4     17              Speed Slammers        1.0
..   ...                         ...        ...
439  715                      Marvel      535.0
440  716                     Modulex        NaN
441  717                 Speed Racer        NaN
442  718       Series 22 Minifigures      535.0
443  719  BrickLink Designer Program        NaN

[444 rows x 3 columns]

In [70]:
df_joined5

set_num                             name  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts     id  version_df  inventory_id     fig_num  quantity  \
0             12   5087           1          5087  fig-000225         1   
1             12   5087           1          5087  fig-000834         1   
2             12   5087           1          5087  fig-004401         1   
3             12   1414           1          1414  fig-000020         2   
4             12   4609           1          4609  fig-001127         2   
...          ...    ...         ...           ...         ...       ...   
18139        147  13813           1         13813  fig-001361         1   
18140          6  34919           1         34919  fig-010738         1   
18141        413  36159           1         36159  fig-004483         1   
18142        413  36159           1         36159  fig-004484         1   
18143        413  36159           1         36159  fig-010456         1   

       version_fig  set_num_fig  
0                1       0011-2  
1                1       0011-2  
2                1       0011-2  
3                1       0012-1  
4                1       0013-1  
...            ...          ...  
18139            1  WEETABIX4-1  
18140            1    WILLIAM-1  
18141            1  YODACHRON-1  
18142            1  YODACHRON-1  
18143            1  YODACHRON-1  

[18144 rows x 12 columns]

In [71]:
def get_minifigs(set_num, version_num):
    df = df_joined6[(df_joined6['set_num'] == set_num)   &  (df_joined6['version_fig'] == version_num)]
    return df[['set_num', 'fig_num', 'fig_name', 'fig_num_parts']].sort_values(by='fig_num').reset_index(drop=True).drop_duplicates()

In [114]:
def minifigs_from_themes(theme_id):
    df = df_joined6[(df_joined6['theme_id'] == theme_id)]
    return df[['fig_num', 'fig_name']].drop_duplicates(subset='fig_num').sort_values(by='fig_num').reset_index(drop=True)

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in `Q16`.  There should be 149 minifigs output for that respective theme.

In [115]:
minifigs_from_themes(171)

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q21:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [74]:
df_joined6

set_num                           name_x  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts_x     id  version_df  inventory_id     fig_num  quantity  \
0               12   5087           1          5087  fig-000225         1   
1               12   5087           1          5087  fig-000834         1   
2               12   5087           1          5087  fig-004401         1   
3               12   1414           1          1414  fig-000020         2   
4               12   4609           1          4609  fig-001127         2   
...            ...    ...         ...           ...         ...       ...   
18139          147  13813           1         13813  fig-001361         1   
18140            6  34919           1         34919  fig-010738         1   
18141          413  36159           1         36159  fig-004483         1   
18142          413  36159           1         36159  fig-004484         1   
18143          413  36159           1         36159  fig-010456         1   

       version_fig  set_num_fig  \
0                1       0011-2   
1                1       0011-2   
2                1       0011-2   
3                1       0012-1   
4                1       0013-1   
...            ...          ...   
18139            1  WEETABIX4-1   
18140            1    WILLIAM-1   
18141            1  YODACHRON-1   
18142            1  YODACHRON-1   
18143            1  YODACHRON-1   

                                                fig_name  fig_num_parts  
0      Policeman, Black Suit with Pocket and Badge, W...              4  
1                Man, Red Torso, Blue Legs, Red Hard Hat              4  
2                           Woman - Blue Torso, Red Legs              4  
3      Classic Spaceman, Red with Airtanks (3842a Hel...              5  
4      Classic Spaceman, White with Airtanks (3842a H...              5  
...                                                  ...            ...  
18139         LEGOLAND White Torso, Black Legs, Red Hair              4  
18140                                          Will.i.am              4  
18141  Yoda, Sand Green Head with Gray Hair, Dark Ora...              3  
18142                                    Padawan, Female              4  
18143                                     Holocron Droid             15  

[18144 rows x 14 columns]

In [75]:
df_joined6 = df_joined6.rename({'name_x': 'name_sets'}, axis=1)
df_joined6

set_num                        name_sets  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts_x     id  version_df  inventory_id     fig_num  quantity  \
0               12   5087           1          5087  fig-000225         1   
1               12   5087           1          5087  fig-000834         1   
2               12   5087           1          5087  fig-004401         1   
3               12   1414           1          1414  fig-000020         2   
4               12   4609           1          4609  fig-001127         2   
...            ...    ...         ...           ...         ...       ...   
18139          147  13813           1         13813  fig-001361         1   
18140            6  34919           1         34919  fig-010738         1   
18141          413  36159           1         36159  fig-004483         1   
18142          413  36159           1         36159  fig-004484         1   
18143          413  36159           1         36159  fig-010456         1   

       version_fig  set_num_fig  \
0                1       0011-2   
1                1       0011-2   
2                1       0011-2   
3                1       0012-1   
4                1       0013-1   
...            ...          ...   
18139            1  WEETABIX4-1   
18140            1    WILLIAM-1   
18141            1  YODACHRON-1   
18142            1  YODACHRON-1   
18143            1  YODACHRON-1   

                                                fig_name  fig_num_parts  
0      Policeman, Black Suit with Pocket and Badge, W...              4  
1                Man, Red Torso, Blue Legs, Red Hard Hat              4  
2                           Woman - Blue Torso, Red Legs              4  
3      Classic Spaceman, Red with Airtanks (3842a Hel...              5  
4      Classic Spaceman, White with Airtanks (3842a H...              5  
...                                                  ...            ...  
18139         LEGOLAND White Torso, Black Legs, Red Hair              4  
18140                                          Will.i.am              4  
18141  Yoda, Sand Green Head with Gray Hair, Dark Ora...              3  
18142                                    Padawan, Female              4  
18143                                     Holocron Droid             15  

[18144 rows x 14 columns]

In [76]:
df_joined6 = df_joined6.rename({'version_df': 'version'}, axis=1)
df_joined6

set_num                        name_sets  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts_x     id  version  inventory_id     fig_num  quantity  \
0               12   5087        1          5087  fig-000225         1   
1               12   5087        1          5087  fig-000834         1   
2               12   5087        1          5087  fig-004401         1   
3               12   1414        1          1414  fig-000020         2   
4               12   4609        1          4609  fig-001127         2   
...            ...    ...      ...           ...         ...       ...   
18139          147  13813        1         13813  fig-001361         1   
18140            6  34919        1         34919  fig-010738         1   
18141          413  36159        1         36159  fig-004483         1   
18142          413  36159        1         36159  fig-004484         1   
18143          413  36159        1         36159  fig-010456         1   

       version_fig  set_num_fig  \
0                1       0011-2   
1                1       0011-2   
2                1       0011-2   
3                1       0012-1   
4                1       0013-1   
...            ...          ...   
18139            1  WEETABIX4-1   
18140            1    WILLIAM-1   
18141            1  YODACHRON-1   
18142            1  YODACHRON-1   
18143            1  YODACHRON-1   

                                                fig_name  fig_num_parts  
0      Policeman, Black Suit with Pocket and Badge, W...              4  
1                Man, Red Torso, Blue Legs, Red Hard Hat              4  
2                           Woman - Blue Torso, Red Legs              4  
3      Classic Spaceman, Red with Airtanks (3842a Hel...              5  
4      Classic Spaceman, White with Airtanks (3842a H...              5  
...                                                  ...            ...  
18139         LEGOLAND White Torso, Black Legs, Red Hair              4  
18140                                          Will.i.am              4  
18141  Yoda, Sand Green Head with Gray Hair, Dark Ora...              3  
18142                                    Padawan, Female              4  
18143                                     Holocron Droid             15  

[18144 rows x 14 columns]

In [77]:
def sets_from_minifig(fig_num):
    df = df_joined6[(df_joined6['fig_num'] == fig_num)]
    return df[[ 'set_num', 'version', 'name_sets', 'fig_num', 'fig_name']].sort_values(by=['set_num', 'version']).reset_index(drop=True)  

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [78]:
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q22:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [79]:
themes

id                        name  parent_id
0      1                     Technic        NaN
1      3                 Competition        1.0
2      4              Expert Builder        1.0
3     16                  RoboRiders        1.0
4     17              Speed Slammers        1.0
..   ...                         ...        ...
439  715                      Marvel      535.0
440  716                     Modulex        NaN
441  717                 Speed Racer        NaN
442  718       Series 22 Minifigures      535.0
443  719  BrickLink Designer Program        NaN

[444 rows x 3 columns]

In [80]:
sets

set_num                             name  year  theme_id  num_parts
0            001-1                            Gears  1965         1         43
1           0011-2                Town Mini-Figures  1979        67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199          0
3           0012-1               Space Mini-Figures  1979       143         12
4           0013-1               Space Mini-Figures  1979       143         12
...            ...                              ...   ...       ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019       158         60
18572      XWING-2                X-Wing Trench Run  2019       158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158        413
18574   YTERRIER-1                Yorkshire Terrier  2018       598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501          0

[18576 rows x 5 columns]

In [81]:
colors

id                  name     rgb is_trans
0      -1             [Unknown]  0033B2        f
1       0                 Black  05131D        f
2       1                  Blue  0055BF        f
3       2                 Green  237841        f
4       3        Dark Turquoise  008F9B        f
..    ...                   ...     ...      ...
185  1059     Trans-Purple Opal  8320B7        t
186  1060      Trans-Green Opal  84B68D        t
187  1061  Trans-Dark Blue Opal  0020A0        t
188  1062        Vibrant Yellow  EBD800        f
189  9999  [No Color/Any Color]  05131D        f

[190 rows x 4 columns]

In [82]:
inventory_parts

inventory_id        part_num  color_id  quantity is_spare
0                  1        48379c01        72         1        f
1                  1           48395         7         1        f
2                  1  stickerupn0077      9999         1        f
3                  1         upn0342         0         1        f
4                  1         upn0350        25         1        f
...              ...             ...       ...       ...      ...
989524        122472           99206        71         2        f
989525        122472           99207        71         2        f
989526        122472           99207         0         2        f
989527        122472           99780         0         2        f
989528        122472           99780         1         1        f

[989529 rows x 5 columns]

In [83]:
inventories

id  version        set_num
0           1        1         7922-1
1           3        1         3931-1
2           4        1         6942-1
3          15        1         5158-1
4          16        1          903-1
...       ...      ...            ...
31582  122328        2          426-1
31583  122329        1  210-2-DUMMY-1
31584  122344        2         7036-1
31585  122450        1        11961-1
31586  122472        3        10262-1

[31587 rows x 3 columns]

In [84]:
parts

part_num                                        name  part_cat_id  \
0           003381                 Sticker Sheet for Set 663-1           58   
1           003383         Sticker Sheet for Sets 618-1, 628-2           58   
2           003402  Sticker Sheet for Sets 310-3, 311-1, 312-3           58   
3           003429                Sticker Sheet for Set 1550-1           58   
4           003432  Sticker Sheet for Sets 357-1, 355-1, 940-1           58   
...            ...                                         ...          ...   
44776  znapupn0002       Znap Connector 3 x 3 - 4 way B (Beam)           43   
44777  znapupn0003              Znap Connector 1 x 3 - 2 way A           43   
44778  znapupn0004                         Znap Beam 3, 1 Hole           43   
44779  znapupn0005     Znap Connector 3 x 3 - 4 way C (Closed)           43   
44780  znapupn0006           Wheel 68mm Znap Propeller (9 x 2)           29   

      part_material  
0           Plastic  
1           Plastic  
2           Plastic  
3           Plastic  
4           Plastic  
...             ...  
44776       Plastic  
44777       Plastic  
44778       Plastic  
44779       Plastic  
44780       Plastic  

[44781 rows x 4 columns]

In [85]:
df_joined7 = sets.merge(inventories, left_on='set_num', right_on='set_num')
df_joined7

set_num                             name  year  theme_id  \
0            001-1                            Gears  1965         1   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
19755      XWING-1              Mini X-Wing Fighter  2019       158   
19756      XWING-2                X-Wing Trench Run  2019       158   
19757  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
19758   YTERRIER-1                Yorkshire Terrier  2018       598   
19759     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501   

       num_parts     id  version  
0             43  24696        1  
1             12   5087        1  
2              0   2216        1  
3             12   1414        1  
4             12   4609        1  
...          ...    ...      ...  
19755         60  35852        1  
19756         52  37243        1  
19757        413  36159        1  
19758          0  88108        1  
19759          0  74972        1  

[19760 rows x 7 columns]

In [86]:
df_joined8 = df_joined7.merge(themes, left_on='theme_id', right_on='id', suffixes=['_sets', '_themes'])
df_joined8

set_num                           name_sets  year  theme_id  \
0            001-1                               Gears  1965         1   
1            002-1      4.5V Samsonite Gears Motor Set  1965         1   
2           1030-1      TECHNIC I: Simple Machines Set  1985         1   
3           1038-1                  ERBIE the Robo-Car  1985         1   
4           1039-1                Manual Control Set 1  1986         1   
...            ...                                 ...   ...       ...   
19755  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963       716   
19756  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963       716   
19757  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963       716   
19758  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963       716   
19759  WEETABIX1-1                     Weetabix Castle  1970       414   

       num_parts  id_sets  version  id_themes name_themes  parent_id  
0             43    24696        1          1     Technic        NaN  
1              3    24701        1          1     Technic        NaN  
2            211    15686        1          1     Technic        NaN  
3            120    13420        1          1     Technic        NaN  
4             39    10042        1          1     Technic        NaN  
...          ...      ...      ...        ...         ...        ...  
19755         10   109713        1        716     Modulex        NaN  
19756         10   109714        2        716     Modulex        NaN  
19757         10   109715        1        716     Modulex        NaN  
19758         10   109716        2        716     Modulex        NaN  
19759        471     5574        1        414      Castle      411.0  

[19760 rows x 10 columns]

 - the DataFrame still needs columns/column names: , part_num, name_part, name_color

In [87]:
colors

id                  name     rgb is_trans
0      -1             [Unknown]  0033B2        f
1       0                 Black  05131D        f
2       1                  Blue  0055BF        f
3       2                 Green  237841        f
4       3        Dark Turquoise  008F9B        f
..    ...                   ...     ...      ...
185  1059     Trans-Purple Opal  8320B7        t
186  1060      Trans-Green Opal  84B68D        t
187  1061  Trans-Dark Blue Opal  0020A0        t
188  1062        Vibrant Yellow  EBD800        f
189  9999  [No Color/Any Color]  05131D        f

[190 rows x 4 columns]

In [88]:
#df_joined9 = df_joined8.merge(colors, left_on='id_sets', right_on='inventory_id')
#df_joined9

In [89]:
#df_joined9 = df_joined8.merge(inventory_parts, left_on='part_num', right_on='part_num')
#df_joined9 = df_joined10.rename({'id_sets': 'id'}, axis=1)
#df_joined9

In [90]:
#def set_parts(set_num, version):
    #df = df_joined6[(df_joined6['set_num'] == set_num)   &  (df_joined6['version_fig'] == version)]
    #return df[['set_num', 'version', 'name_sets', 'year', 'name_themes']].sort_values(by='fig_num').reset_index(drop=True).drop_duplicates()

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in `Q16`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [91]:
### CODE CHECK ###

  **Q23:**  Get a list of all unique color names, sort by color name, and output as a list.  Save this as `Q23`. 

In [92]:
#def main2():
    #for i in colors.name.unique():
       # mylist = colors.name.unique()
       # mylist2 = np.sort(mylist)
        #return mylist2 

In [93]:
#main2()

In [94]:
for i in colors.name.unique():
    mylist = colors.name.unique()
    mylist2 = np.sort(mylist)
    print(mylist2)

['Aqua' 'Black' 'Blue' 'Blue-Violet' 'Bright Green' 'Bright Light Blue'
 'Bright Light Orange' 'Bright Light Yellow' 'Bright Pink' 'Brown'
 'Chrome Antique Brass' 'Chrome Black' 'Chrome Blue' 'Chrome Gold'
 'Chrome Green' 'Chrome Pink' 'Chrome Silver' 'Copper' 'Coral'
 'Dark Azure' 'Dark Blue' 'Dark Blue-Violet' 'Dark Bluish Gray'
 'Dark Brown' 'Dark Flesh' 'Dark Gray' 'Dark Green' 'Dark Orange'
 'Dark Pink' 'Dark Purple' 'Dark Red' 'Dark Tan' 'Dark Turquoise'
 'Earth Orange' 'Fabuland Brown' 'Fabuland Orange' 'Flat Dark Gold'
 'Flat Silver' 'Flesh' 'Glitter Trans-Clear' 'Glitter Trans-Dark Pink'
 'Glitter Trans-Light Blue' 'Glitter Trans-Neon Green'
 'Glitter Trans-Orange' 'Glitter Trans-Purple' 'Glow In Dark Opaque'
 'Glow In Dark Trans' 'Glow in Dark White' 'Green' 'Lavender' 'Light Aqua'
 'Light Blue' 'Light Bluish Gray' 'Light Flesh' 'Light Gray' 'Light Green'
 'Light Lime' 'Light Orange' 'Light Pink' 'Light Purple' 'Light Salmon'
 'Light Turquoise' 'Light Violet' 'Light Yellow' '

In [95]:
Q23 = mylist2.tolist()
Q23

['Aqua',
 'Black',
 'Blue',
 'Blue-Violet',
 'Bright Green',
 'Bright Light Blue',
 'Bright Light Orange',
 'Bright Light Yellow',
 'Bright Pink',
 'Brown',
 'Chrome Antique Brass',
 'Chrome Black',
 'Chrome Blue',
 'Chrome Gold',
 'Chrome Green',
 'Chrome Pink',
 'Chrome Silver',
 'Copper',
 'Coral',
 'Dark Azure',
 'Dark Blue',
 'Dark Blue-Violet',
 'Dark Bluish Gray',
 'Dark Brown',
 'Dark Flesh',
 'Dark Gray',
 'Dark Green',
 'Dark Orange',
 'Dark Pink',
 'Dark Purple',
 'Dark Red',
 'Dark Tan',
 'Dark Turquoise',
 'Earth Orange',
 'Fabuland Brown',
 'Fabuland Orange',
 'Flat Dark Gold',
 'Flat Silver',
 'Flesh',
 'Glitter Trans-Clear',
 'Glitter Trans-Dark Pink',
 'Glitter Trans-Light Blue',
 'Glitter Trans-Neon Green',
 'Glitter Trans-Orange',
 'Glitter Trans-Purple',
 'Glow In Dark Opaque',
 'Glow In Dark Trans',
 'Glow in Dark White',
 'Green',
 'Lavender',
 'Light Aqua',
 'Light Blue',
 'Light Bluish Gray',
 'Light Flesh',
 'Light Gray',
 'Light Green',
 'Light Lime',
 'Light 

**Q24:** How many total transparent colors are in the `colors` DataFrame?  Save this as `Q24` (as an integer).

In [96]:
mystring = str(Q23)
mystring

"['Aqua', 'Black', 'Blue', 'Blue-Violet', 'Bright Green', 'Bright Light Blue', 'Bright Light Orange', 'Bright Light Yellow', 'Bright Pink', 'Brown', 'Chrome Antique Brass', 'Chrome Black', 'Chrome Blue', 'Chrome Gold', 'Chrome Green', 'Chrome Pink', 'Chrome Silver', 'Copper', 'Coral', 'Dark Azure', 'Dark Blue', 'Dark Blue-Violet', 'Dark Bluish Gray', 'Dark Brown', 'Dark Flesh', 'Dark Gray', 'Dark Green', 'Dark Orange', 'Dark Pink', 'Dark Purple', 'Dark Red', 'Dark Tan', 'Dark Turquoise', 'Earth Orange', 'Fabuland Brown', 'Fabuland Orange', 'Flat Dark Gold', 'Flat Silver', 'Flesh', 'Glitter Trans-Clear', 'Glitter Trans-Dark Pink', 'Glitter Trans-Light Blue', 'Glitter Trans-Neon Green', 'Glitter Trans-Orange', 'Glitter Trans-Purple', 'Glow In Dark Opaque', 'Glow In Dark Trans', 'Glow in Dark White', 'Green', 'Lavender', 'Light Aqua', 'Light Blue', 'Light Bluish Gray', 'Light Flesh', 'Light Gray', 'Light Green', 'Light Lime', 'Light Orange', 'Light Pink', 'Light Purple', 'Light Salmon', '

In [97]:
Q24 = mystring.count('Trans')
Q24

38

**Q25:**  
- Create a function called `get_part_colors` that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.  
- The DataFrame should be sorted by color id  
- The index should go from 0 to n-1.

In [98]:
elements.head()

element_id  part_num  color_id
0    6300211  67906c01        14
1    4566309      2564         0
2    4275423     53657      1004
3    6194308     92926        71
4    6229123     26561         4

In [99]:
colors.head()

id            name     rgb is_trans
0  -1       [Unknown]  0033B2        f
1   0           Black  05131D        f
2   1            Blue  0055BF        f
3   2           Green  237841        f
4   3  Dark Turquoise  008F9B        f

In [100]:
ElCol_df = pd.merge(elements, colors, left_on='color_id', right_on='id')
ElCol_df

element_id     part_num  color_id    id               name     rgb  \
0        6300211     67906c01        14    14             Yellow  F2CD37   
1        6284189        32072        14    14             Yellow  F2CD37   
2        6323851        11938        14    14             Yellow  F2CD37   
3         246524         2465        14    14             Yellow  F2CD37   
4        4501532     57912c01        14    14             Yellow  F2CD37   
...          ...          ...       ...   ...                ...     ...   
62679    6323781       44375b      1056  1056   Trans-Brown Opal  583927   
62680    6288478        64462      1059  1059  Trans-Purple Opal  8320B7   
62681    6172108  22385pr0040        54    54  Trans-Neon Yellow  DAB000   
62682    4107024         6969       100   100       Light Salmon  FEBABD   
62683    6341650  3068bpr9777      1025  1025     Modulex Orange  F47B30   

      is_trans  
0            f  
1            f  
2            f  
3            f  
4            f  
...        ...  
62679        t  
62680        t  
62681        t  
62682        f  
62683        f  

[62684 rows x 7 columns]

In [112]:
def get_part_colors(part_num):
    filtered_df = ElCol_df[ElCol_df['part_num'] == part_num]
    return filtered_df[['id', 'name', 'rgb', 'is_trans']].drop_duplicates().sort_values(by='id').reset_index(drop=True)

In [113]:
#checking
get_part_colors('67906c01')

id       name     rgb is_trans
0   14     Yellow  F2CD37        f
1  272  Dark Blue  0A3463        f

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets?
- What color is used the most for all parts?
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
